docs/dev/sql-nested-function-select-clause.md

```diff
@@ -0,0 +1,484 @@
+    participant DefaultImplementor
+
+%% Parsing
+SQLService->>+ParserBaseRoot:visitRoot
+  ParserBaseRoot->>+AstExpressionBuilder:visitScalarFunction
+  AstExpressionBuilder-->>-ParserBaseRoot:Function
+ParserBaseRoot-->>-SQLService:UnresolvedPlan
+%% Analysis
+SQLService->>+QueryService:analyze
+  QueryService->>+Analyzer:visitProject
+    Analyzer->>+NestedAnalyzer:visitFunction
+    NestedAnalyzer-->>-Analyzer:LogicalNested
+  Analyzer-->>-QueryService:UnresolvedPlan
+    
+  %% planner optimization
+  QueryService->>+Planner:plan
+    Planner->>+TableScanPushDown:apply
+      TableScanPushDown->>+OpenSearchRequestBuilder:pushDownNested
+
+      Note over TableScanPushDown, OpenSearchRequestBuilder: returns false keeping<br>LogicalNested in plan tree
+
+      OpenSearchRequestBuilder-->>-TableScanPushDown:boolean
+    TableScanPushDown-->>-Planner:LogicalPlan
+    %% planner implementation
+    Planner->>+DefaultImplementor:visitNested
+    DefaultImplementor-->>-Planner:NestedOperator
+  Planner-->>-QueryService:PhysicalPlan
+QueryService-->>-SQLService:PhysicalPlan
+```
+
+### 2.3 Sequence Diagram for Nested SELECT Clause Post-processing
+Nested function sequence diagram illustrating the flattening of the OpenSearch response. Flattening the response from OpenSearch changes the nested types structure by making the full path of an object the key, and the object it refers to the value. As well when a user selects multiple nested fields with differing path values, a cross join is done on the result. These examples show the flattening output keys and cross join.
+
+**Sample input:**
+```json
+{
+  "comments": {
+    "data": "abc"
+  },
+  "message": [
+    { "info": "letter1" },
+    { "info": "letter2" }
+  ]
+}
+```
+**Sample Output:**
+```json
+[
+  [
+    { "comment.data": "abc" },
+    { "message.info": "letter1" }
+  ],
+  [
+    { "comment.data": "abc" },
+    { "message.info": "letter2" }
+  ]
+]
+```
+
+```mermaid
+sequenceDiagram
+
+%% Flattening
+OpenSearchExecutionEngine->>+ProjectOperator:next
+  ProjectOperator->>+ResourceMonitorPlan:next
+    ResourceMonitorPlan->>+NestedOperator:next
+      loop unnesting
+        NestedOperator->>NestedOperator:flatten
+      end
+    NestedOperator-->>-ResourceMonitorPlan:ExprValue
+  ResourceMonitorPlan-->>-ProjectOperator:ExprValue
+ProjectOperator-->>-OpenSearchExecutionEngine:ExprValue
+```
+
+### 2.4 Select Clause Nested Query Class Diagram
+Nested function class diagram for additional classes required for query execution. The `NestedAnalyzer` is a visitor for nested functions used in the SELECT clause to fulfill the `LogicalNested` LogicalPlan. After push down is successful the `NestedOperator` PhysicalPlan is used for object flattening of the OpenSearch response.
+
+```mermaid
+---
+title: Nested Select Clause Class Diagram
+---
+classDiagram
+
+    NestedAnalyzer~AbstractNodeVisitor~-->LogicalNested~LogicalPlan~
+    LogicalNested-->NestedOperator~PhysicalPlan~
```

Comment:
What do you want to express with this graph?

`NestedAnalyzer` creates a `LogicalNested`, so  that association makes sense, but there is no static link between `LogicalNested` and `NestedOperator`.

Reply:
Maybe include the class that creates it?
IntelliJ creates something like this

```mermaid
classDiagram
direction BT
class DefaultImplementor {
  + visitNested(LogicalNested, C) PhysicalPlan

}
class LogicalNested {
  + getProjectList() List~NamedExpression~
  + accept(LogicalPlanNodeVisitor, C) R
  + addFields(Map~String_ReferenceExpression~) void
}
class NestedAnalyzer {
  + analyze(UnresolvedExpression, AnalysisContext) LogicalPlan
  + visitFunction(Function, AnalysisContext) LogicalPlan
  + visitAlias(Alias, AnalysisContext) LogicalPlan
}
class NestedOperator

DefaultImplementor~C~  ..>  NestedOperator : «create»
NestedAnalyzer  ..>  LogicalNested : «create»


```